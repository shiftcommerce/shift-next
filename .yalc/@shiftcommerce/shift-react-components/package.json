{
  "name": "@shiftcommerce/shift-react-components",
  "version": "0.4.0-3362f14d",
  "main": "dist/bundle.js",
  "license": "MIT",
  "scripts": {
    "start": "NODE_ENV=production ANALYZE_BUNDLE=true ./node_modules/webpack/bin/webpack.js --watch --color -p",
    "build": "NODE_ENV=production ./node_modules/webpack/bin/webpack.js --bail -p",
    "build:develop": "NODE_ENV=development ./node_modules/webpack/bin/webpack.js --bail --color -d",
    "test": "jest",
    "lint:scss": "stylelint src/scss",
    "lint:scss:fix": "stylelint src/scss --fix"
  },
  "dependencies": {
    "classnames": "^2.2.6",
    "date-fns": "^1.29.0",
    "formik": "^1.4.0",
    "moment-business-days": "^1.1.3",
    "react": "^16.7.0",
    "react-dom": "^16.7.0",
    "react-input-range": "^1.3.0",
    "react-instantsearch": "^5.3.1",
    "react-instantsearch-dom": "^5.3.1",
    "react-lazyload-fadein": "^1.1.0",
    "react-loading": "^2.0.3",
    "react-pluralize": "^1.5.0",
    "react-redux": "^5.0.7",
    "react-responsive-carousel": "^3.1.43",
    "react-stickyfill": "^0.2.3",
    "react-stripe-elements": "^2.0.3",
    "redux": "^3.7.2",
    "typy": "^2.0.1",
    "yup": "^0.26.6"
  },
  "browserslist": [
    "Last 4 Chrome versions",
    "Last 4 Firefox versions",
    "IE >= 11",
    "Last 3 Safari versions",
    "last 2 iOS versions",
    "Last 2 Android versions",
    "Last 2 ChromeAndroid versions",
    "> 2%"
  ]
}
