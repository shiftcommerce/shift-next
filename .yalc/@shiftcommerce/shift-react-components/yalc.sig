3362f14d59536b855f88913b9280cf38