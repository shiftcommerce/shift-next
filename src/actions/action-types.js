// Action types
export const AUTOFILL_ADDRESS = 'AUTOFILL_ADDRESS'
export const AUTOFILL_BILLING_ADDRESS = 'AUTOFILL_BILLING_ADDRESS'
export const BILLING_ADDRESS_CREATED = 'BILLING_ADDRESS_CREATED'
export const CARD_ERRORS = 'CARD_ERRORS'
export const CART_UPDATED = 'CART_UPDATED'
export const CREATE_ACCOUNT = 'CREATE_ACCOUNT'
export const CREATE_ORDER = 'CREATE_ORDER'
export const ERROR_ACCOUNT = 'ERROR_ACCOUNT'
export const ERROR_LOGIN = 'ERROR_LOGIN'
export const ERROR_MENU = 'ERROR_MENU'
export const ERROR_ORDER = 'ERROR_ORDER'
export const ERROR_REGISTRATION = 'ERROR_REGISTRATION'
export const CLEAR_ACCOUNT_ERRORS = 'CLEAR_ACCOUNT_ERRORS'
export const GET_ACCOUNT = 'GET_ACCOUNT'
export const GET_LOGIN = 'GET_LOGIN'
export const GET_MENU = 'GET_MENU'
export const GET_PRODUCT = 'GET_PRODUCT'
export const GET_CUSTOMER_ORDERS = 'GET_CUSTOMER_ORDERS'
export const PASSWORD_RESET = 'PASSWORD_RESET'
export const REQUEST_CARD_TOKEN = 'REQUEST_CARD_TOKEN'
export const SET_PRODUCT = 'SET_PRODUCT'
export const SET_ACCOUNT = 'SET_ACCOUNT'
export const SET_LOGGED_IN = 'SET_LOGGED_IN'
export const SET_LOGIN = 'SET_LOGIN'
export const SET_CUSTOMER_ORDERS = 'SET_CUSTOMER_ORDERS'
export const SET_ADDRESS = 'SET_ADDRESS'
export const SET_ADDRESS_BOOK_ENTRY_BILLING = 'SET_ADDRESS_BOOK_ENTRY_BILLING'
export const SET_ADDRESS_BOOK_ENTRY_SHIPPING = 'SET_ADDRESS_BOOK_ENTRY_SHIPPING'
export const SET_ADDRESS_BOOK = 'SET_ADDRESS_BOOK'
export const SET_CARD_TOKEN = 'SET_CARD_TOKEN'
export const SET_ORDER = 'SET_ORDER'
export const DELETE_ADDRESS = 'DELETE_ADDRESS'
export const SHIPPING_ADDRESS_CREATED = 'SHIPPING_ADDRESS_CREATED'
export const SET_CHECKOUT_INPUT_VALUE = 'SET_CHECKOUT_INPUT_VALUE'
export const SET_CHECKOUT_INPUT_VALIDATION_MESSAGE = 'SET_CHECKOUT_INPUT_VALIDATION_MESSAGE'
export const SET_CHECKOUT_INPUT_SHOWN = 'SET_CHECKOUT_INPUT_SHOWN'
export const SET_MENU = 'SET_MENU'
export const SET_PAYMENT_ERROR = 'SET_PAYMENT_ERROR'
