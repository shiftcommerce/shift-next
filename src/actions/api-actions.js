// Libs
import ApiClient from '../lib/api-client'
import JsonApiParser from '../lib/json-api-parser'

export const readEndpoint = (request) => {
  return (dispatch) => {
    dispatchInitialAction(dispatch, request)
    return new ApiClient().read(request.endpoint, request.query)
      .then(processResponse(dispatch, request, [200, 304]))
      .catch((error) => {
        if (request.errorActionType) dispatch(setErroredTo(request.errorActionType, error, request))
      })
  }
}

export const postEndpoint = (request) => {
  return (dispatch) => {
    dispatchInitialAction(dispatch, request)
    return new ApiClient().post(request.endpoint, request.body)
      .then(response => {
        _determinePostDispatch(dispatch, request, response)
      })
      .catch(error => {
        if (request.errorActionType) dispatch(setErroredTo(request.errorActionType, error))
      })
  }
}

export const deleteEndpoint = (request) => {
  return (dispatch) => {
    dispatchInitialAction(dispatch, request)
    return new ApiClient().delete(request.endpoint, request.body)
      .then(processResponse(dispatch, request, [204]))
      .catch(error => {
        if (request.errorActionType) dispatch(setErroredTo(request.errorActionType, error))
      })
  }
}

function dispatchInitialAction (dispatch, request) {
  if (request.requestActionType) {
    dispatch(fetching(request.requestActionType, request.requestActionData))
  }
}

function sendResponse (actionType, parsedPayload) {
  return {
    type: actionType,
    payload: parsedPayload
  }
}

function fetching (actionType, actionData) {
  return {
    type: actionType,
    data: actionData
  }
}

function setErroredTo (actionType, data, request) {
  return {
    type: actionType,
    payload: {
      error: {
        data: data,
        request: request
      }
    }
  }
}

function _checkForErrors (response) {
  let validationPassed = true
  // Validation for customer account form
  if (response.data.length > 0) {
    response.data.forEach(item => {
      if (parseInt(item.status) === 422 || parseInt(item.status) === 404) {
        validationPassed = false
      }
    })
  }

  return validationPassed
}

function processResponse (dispatch, request, expectedStatusCodes) {
  return (response) => {
    if (expectedStatusCodes.includes(response.status)) {
      if (request.successActionType) {
        const parsedPayload = new JsonApiParser().parse(response.data)
        // TODO: remove this when all endpoints have been extracted into shift-api

        // If parsedPayload is empty or is an array of objects with only one key (id) it means it had already been parsed.
        let payload
        if (!parsedPayload || (Array.isArray(parsedPayload.data) && parsedPayload.data.every(e => Object.keys(e).length === 1))) {
          payload = response.data
        } else {
          payload = parsedPayload
        }
        dispatch(sendResponse(request.successActionType, payload))
      }
    } else {
      if (request.errorActionType) dispatch(setErroredTo(request.errorActionType, response.data, request))
    }
  }
}

function _determinePostDispatch (dispatch, request, response) {
  const validationPassed = _checkForErrors(response)

  if ((response.status === 202 || response.status === 201 || response.status === 200) && validationPassed) {
    if (request.successActionType) {
      const parsedPayload = new JsonApiParser().parse(response.data)
      // TODO: remove when all enpoints have been moved to shift-next
      const payload = parsedPayload || response.data
      dispatch(sendResponse(request.successActionType, payload))
    }
  } else {
    dispatch(setErroredTo(request.errorActionType, response.data))
  }
}
