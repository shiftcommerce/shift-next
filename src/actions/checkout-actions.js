// Actions
import * as actionTypes from './action-types'

// Store the input change info in the local redux store
function storeInputChange (formName, fieldName, fieldValue) {
  return {
    type: actionTypes.SET_CHECKOUT_INPUT_VALUE,
    payload: {
      formName: formName,
      fieldName: fieldName,
      fieldValue: fieldValue
    }
  }
}

function storeValidationMessage (formName, fieldName, validationMessage) {
  return {
    type: actionTypes.SET_CHECKOUT_INPUT_VALIDATION_MESSAGE,
    payload: {
      formName: formName,
      fieldName: fieldName,
      validationMessage: validationMessage
    }
  }
}

export function autoFillAddress (address) {
  return {
    type: actionTypes.AUTOFILL_ADDRESS,
    address: address
  }
}

export function autoFillBillingAddress (address) {
  return {
    type: actionTypes.AUTOFILL_BILLING_ADDRESS,
    address: address
  }
}

export function inputChange (formName, fieldName, fieldValue) {
  return (dispatch) => {
    dispatch(storeInputChange(formName, fieldName, fieldValue))
  }
}

export function setValidationMessage (formName, fieldName, errorMessage) {
  return (dispatch) => {
    dispatch(storeValidationMessage(formName, fieldName, errorMessage))
  }
}

export function showField (formName, fieldName) {
  return {
    type: actionTypes.SET_CHECKOUT_INPUT_SHOWN,
    payload: {
      formName: formName,
      fieldName: fieldName
    }
  }
}
