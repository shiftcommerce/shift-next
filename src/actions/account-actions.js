// actionTypes
import * as types from '../actions/action-types'

// Actions
import { readEndpoint } from './api-actions'

const accountRequest = {
  endpoint: '/getAccount',
  query: {},
  errorActionType: types.ERROR_ACCOUNT,
  requestActionType: types.GET_ACCOUNT,
  successActionType: types.SET_ACCOUNT
}

export function fetchAccountDetails (store) {
  return readEndpoint(accountRequest)
}

export function clearErrors () {
  return {
    type: types.CLEAR_ACCOUNT_ERRORS
  }
}

export function setLoggedInFromCookies () {
  return {
    type: types.SET_LOGGED_IN,
    payload: {}
  }
}

const customerOrdersRequest = {
  endpoint: `/customerOrders`,
  requestActionType: types.GET_CUSTOMER_ORDERS,
  successActionType: types.SET_CUSTOMER_ORDERS
}

export function getCustomerOrders () {
  return readEndpoint(customerOrdersRequest)
}
