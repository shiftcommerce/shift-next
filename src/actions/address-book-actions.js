// actionTypes
import * as actionTypes from './action-types'

// Actions
import { readEndpoint, postEndpoint, deleteEndpoint } from './api-actions'

export function saveToAddressBook (address, options = {}) {
  const request = {
    endpoint: '/createAddressBookEntry',
    body: newAddressBookEntryPayload(address),
    successActionType: options.billing ? actionTypes.SET_ADDRESS_BOOK_ENTRY_BILLING : actionTypes.SET_ADDRESS_BOOK_ENTRY_SHIPPING
  }
  return postEndpoint(request)
}

export function fetchAddressBook () {
  const request = {
    endpoint: '/getAddressBook',
    successActionType: actionTypes.SET_ADDRESS_BOOK
  }
  return readEndpoint(request)
}

export function deleteAddressBookEntry (address) {
  const request = {
    endpoint: `/deleteAddress/${address.id}`,
    requestActionType: actionTypes.DELETE_ADDRESS,
    requestActionData: {
      addressId: address.id
    }
  }
  return deleteEndpoint(request)
}

export function setAddress (address, formName) {
  return {
    type: actionTypes.SET_ADDRESS,
    address: address,
    formName: formName
  }
}

function newAddressBookEntryPayload (address) {
  return {
    data: {
      type: 'addresses',
      attributes: {
        first_name: address.first_name,
        last_name: address.last_name,
        address_line_1: address.line_1,
        address_line_2: address.line_2,
        city: address.city,
        state: address.state,
        country: address.country_code,
        postcode: address.zipcode,
        preferred_billing: address.preferred_billing,
        preferred_shipping: address.preferred_shipping,
        meta_attributes: addressMetaAttributes(address)
      }
    }
  }
}

function addressMetaAttributes (address) {
  return {
    label: {
      value: address.label
    },
    company_name: {
      value: address.companyName
    },
    phone_number: {
      value: address.primary_phone
    },
    email: {
      value: address.email
    }
  }
}
