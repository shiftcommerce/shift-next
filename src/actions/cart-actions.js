// Libraries
import Cookies from 'js-cookie'

// Lib
import ApiClient from '../lib/api-client'

// Actions
import * as actionTypes from './action-types'
import { readEndpoint, postEndpoint } from './api-actions'

const fetchCartRequest = {
  endpoint: '/getCart',
  successActionType: actionTypes.CART_UPDATED
}

export function readCart (options = {}) {
  return (dispatch, getState) => {
    // Only fetch the cart if needed, otherwise do nothing
    if ((Cookies.get('cart') && !getState().cart.id) || options.force) {
      return dispatch(readEndpoint(fetchCartRequest))
    } else {
      return Promise.resolve()
    }
  }
}

const addToCartRequest = (variantId, quantity) => {
  return {
    endpoint: '/addToCart',
    body: { variantId, quantity },
    successActionType: actionTypes.CART_UPDATED
  }
}

export function addToCart (variantId, quantity) {
  return postEndpoint(addToCartRequest(variantId, quantity))
}

const updateLineItemQuantityRequest = (lineItemId, newQuantity) => {
  return {
    endpoint: '/updateLineItem',
    body: { lineItemId, newQuantity },
    successActionType: actionTypes.CART_UPDATED
  }
}

export function updateLineItemQuantity (lineItemId, newQuantity) {
  return postEndpoint(updateLineItemQuantityRequest(lineItemId, newQuantity))
}

const deleteLineItemRequest = (lineItemId) => {
  return {
    endpoint: `/deleteLineItem/${lineItemId}`,
    successActionType: actionTypes.CART_UPDATED
  }
}

export function deleteLineItem (lineItemId) {
  return postEndpoint(deleteLineItemRequest(lineItemId))
}

const addCartCouponRequest = (couponCode) => {
  return {
    endpoint: '/addCartCoupon',
    body: { couponCode }
  }
}

export const fetchShippingMethodsRequest = () => {
  return {
    endpoint: '/getShippingMethods'
  }
}

const setShippingMethodRequest = (shippingMethodId) => {
  return {
    endpoint: '/setShippingMethod',
    body: { shippingMethodId },
    successActionType: actionTypes.CART_UPDATED
  }
}

export function setCartShippingMethod (shippingMethodId) {
  return postEndpoint(setShippingMethodRequest(shippingMethodId))
}

export async function submitCoupon (couponCode) {
  const request = addCartCouponRequest(couponCode)
  const response = await new ApiClient().post(request.endpoint, request.body)
  if (response.status === 201) {
    return response.data
  }
  throw response.data
}

export function setAPIError (error, setErrors) {
  setErrors({ couponCode: error[0]['title'] })
}

const setBillingAddressRequest = (addressId) => {
  return {
    endpoint: '/setCartBillingAddress',
    body: { addressId },
    successActionType: actionTypes.CART_UPDATED
  }
}

export function setCartBillingAddress (addressId) {
  return postEndpoint(setBillingAddressRequest(addressId))
}

const setShippingAddressRequest = (addressId) => {
  return {
    endpoint: '/setCartShippingAddress',
    body: { addressId },
    successActionType: actionTypes.CART_UPDATED
  }
}

export function setCartShippingAddress (addressId) {
  return postEndpoint(setShippingAddressRequest(addressId))
}

const createShippingAddressRequest = (address) => {
  return {
    endpoint: '/createAddress',
    body: address,
    successActionType: actionTypes.SHIPPING_ADDRESS_CREATED
  }
}

export function createShippingAddress (address) {
  return postEndpoint(createShippingAddressRequest(address))
}

const createBillingAddressRequest = (address) => {
  return {
    endpoint: '/createAddress',
    body: address,
    successActionType: actionTypes.BILLING_ADDRESS_CREATED
  }
}

export function createBillingAddress (address) {
  return postEndpoint(createBillingAddressRequest(address))
}

export function toggleMiniBag (displayed) {
  return {
    type: actionTypes.TOGGLE_MINIBAG,
    displayed
  }
}
