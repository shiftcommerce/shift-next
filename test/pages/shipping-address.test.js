// Libraries
import Cookies from 'js-cookie'
import Router from 'next/router'

// Pages
import ShippingAddressPage from '../../src/pages/shipping-address'

// Components
import { AddressFormHeader, Button, CheckoutAddressForm } from '@shiftcommerce/shift-react-components'

// Actions
import * as AddressBookActions from '../../src/actions/address-book-actions'
import * as CartActions from '../../src/actions/cart-actions'

jest.mock('next/config', () => () => ({
  publicRuntimeConfig: {}
}))

describe('componentDidMount()', () => {
  test('sets loading to false in state when user is not logged in', () => {
    const wrapper = shallow(<ShippingAddressPage cart={{}} checkout={{}} />)

    wrapper.instance().componentDidMount()

    expect(wrapper.instance().state.loading).toBe(false)
  })

  describe('when user is logged in', () => {
    test("fetches the address book and preselects the preferred shipping address when cart doesn't have a shipping address", async () => {
      const cookieSpy = jest.spyOn(Cookies, 'get').mockImplementation(() => true)
      const fetchAddressBookSpy = jest.spyOn(AddressBookActions, 'fetchAddressBook').mockImplementation(() => 'fetchAddressBookAction')
      const setCartShippingAddressSpy = jest.spyOn(CartActions, 'setCartShippingAddress').mockImplementation(() => 'setCartShippingAddressAction')
      const dispatch = jest.fn().mockImplementation(value => Promise.resolve(value))
      const checkout = {
        addressBook: [{
          id: 1
        }, {
          id: 2,
          preferred_shipping: true
        }],
        paymentMethod: 'PayPal'
      }
      const thirdPartyPaymentMethods = ['PayPal']

      const wrapper = shallow(<ShippingAddressPage cart={{}} dispatch={dispatch} checkout={checkout} thirdPartyPaymentMethods={thirdPartyPaymentMethods} />)

      await wrapper.instance().componentDidMount()

      expect(dispatch).toHaveBeenCalledWith('fetchAddressBookAction')
      expect(setCartShippingAddressSpy).toHaveBeenCalledWith(2)
      expect(dispatch).toHaveBeenCalledWith('setCartShippingAddressAction')
      expect(wrapper.instance().state.loading).toBe(false)

      cookieSpy.mockRestore()
      fetchAddressBookSpy.mockRestore()
      setCartShippingAddressSpy.mockRestore()
    })

    test("fetches the address book and preselects the first shipping address when cart doesn't have a shipping address and there isn't a preferred one", async () => {
      const cookieSpy = jest.spyOn(Cookies, 'get').mockImplementation(() => true)
      const fetchAddressBookSpy = jest.spyOn(AddressBookActions, 'fetchAddressBook').mockImplementation(() => 'fetchAddressBookAction')
      const setCartShippingAddressSpy = jest.spyOn(CartActions, 'setCartShippingAddress').mockImplementation(() => 'setCartShippingAddressAction')
      const dispatch = jest.fn().mockImplementation(value => Promise.resolve(value))
      const checkout = {
        addressBook: [{
          id: 1
        }, {
          id: 2
        }]
      }

      const wrapper = shallow(<ShippingAddressPage cart={{}} dispatch={dispatch} checkout={checkout} />)

      await wrapper.instance().componentDidMount()

      expect(dispatch).toHaveBeenCalledWith('fetchAddressBookAction')
      expect(setCartShippingAddressSpy).toHaveBeenCalledWith(1)
      expect(dispatch).toHaveBeenCalledWith('setCartShippingAddressAction')
      expect(wrapper.instance().state.loading).toBe(false)

      cookieSpy.mockRestore()
      fetchAddressBookSpy.mockRestore()
      setCartShippingAddressSpy.mockRestore()
    })

    test('fetches the address book and sets loading to false when cart has a shipping address', async () => {
      // Arrange
      const cookieSpy = jest.spyOn(Cookies, 'get').mockImplementation(() => true)
      const fetchAddressBookSpy = jest.spyOn(AddressBookActions, 'fetchAddressBook').mockImplementation(() => 'fetchAddressBookAction')
      const dispatch = jest.fn().mockImplementation(value => Promise.resolve(value))
      const cart = { shipping_address: {} }
      const checkoutState = {paymentMethod: 'Credit/Debit Card'}
      const thirdPartyPaymentMethodOptions = ['PayPal']

      // Act
      const wrapper = shallow(<ShippingAddressPage cart={cart} checkout={checkoutState} thirdPartyPaymentMethods={thirdPartyPaymentMethodOptions} dispatch={dispatch} />)
      await wrapper.instance().componentDidMount()

      // Assert
      expect(dispatch).toHaveBeenCalledWith('fetchAddressBookAction')
      expect(wrapper.instance().state.loading).toBe(false)

      cookieSpy.mockRestore()
      fetchAddressBookSpy.mockRestore()
    })
  })
})

describe('nextSection()', () => {
  test('navigates to shipping method selection when shipping address is from the address book', async () => {
    // Arrange
    const cookieSpy = jest.spyOn(Cookies, 'get').mockImplementation(() => true)
    const pushSpy = jest.spyOn(Router, 'push').mockImplementation(() => {})
    const cart = {
      shipping_address: {
        id: 10
      }
    }
    const checkoutState = {
      addressBook: [{
        id: 10
      }],
      paymentMethod: 'Credit/Debit Card'
    }
    const thirdPartyPaymentMethodOptions = ['PayPal']

    // Act
    const wrapper = shallow(<ShippingAddressPage cart={cart} checkout={checkoutState} thirdPartyPaymentMethods={thirdPartyPaymentMethodOptions}  />, { disableLifecycleMethods: true })
    await wrapper.instance().nextSection()

    // Assert
    expect(pushSpy).toHaveBeenCalledWith('/checkout/shipping-method')

    cookieSpy.mockRestore()
    pushSpy.mockRestore()
  })

  test('saves new address to address book and sets it on cart', async () => {
    const cookieSpy = jest.spyOn(Cookies, 'get').mockImplementation(() => true)
    const pushSpy = jest.spyOn(Router, 'push').mockImplementation(() => {})
    const saveToAddressBookSpy = jest.spyOn(AddressBookActions, 'saveToAddressBook').mockImplementation(() => 'saveToAddressBookAction')
    const setCartShippingAddressSpy = jest.spyOn(CartActions, 'setCartShippingAddress').mockImplementation(() => 'setCartShippingAddressAction')
    const dispatch = jest.fn().mockImplementation(() => Promise.resolve())
    const cart = {
      shipping_address: {
        id: 20
      }
    }
    const checkout = {
      shippingAddress: {
        id: 20,
        saveToAddressBook: true
      }
    }

    const wrapper = shallow(<ShippingAddressPage cart={cart} checkout={checkout} dispatch={dispatch} />, { disableLifecycleMethods: true })
    wrapper.setState({ addingNewAddress: true })

    await wrapper.instance().nextSection()

    expect(saveToAddressBookSpy).toHaveBeenCalledWith(checkout.shippingAddress)
    expect(setCartShippingAddressSpy).toHaveBeenCalledWith(20)
    expect(dispatch).toHaveBeenCalledWith('saveToAddressBookAction')
    expect(dispatch).toHaveBeenCalledWith('setCartShippingAddressAction')
    expect(pushSpy).toHaveBeenCalledWith('/checkout/shipping-method')

    cookieSpy.mockRestore()
    pushSpy.mockRestore()
    saveToAddressBookSpy.mockRestore()
    setCartShippingAddressSpy.mockRestore()
  })

  test('creates new addresses and sets them on cart', async () => {
    const cookieSpy = jest.spyOn(Cookies, 'get').mockImplementation(() => true)
    const pushSpy = jest.spyOn(Router, 'push').mockImplementation(() => {})
    const createShippingAddressSpy = jest.spyOn(CartActions, 'createShippingAddress').mockImplementation(() => 'createShippingAddressAction')
    const setCartShippingAddressSpy = jest.spyOn(CartActions, 'setCartShippingAddress').mockImplementation(() => 'setCartShippingAddressAction')
    const dispatch = jest.fn().mockImplementation(() => Promise.resolve())
    const cart = {
      shipping_address: {
        id: 20
      }
    }
    const checkout = {
      shippingAddress: {
        id: 20
      }
    }

    const wrapper = shallow(<ShippingAddressPage cart={cart} checkout={checkout} dispatch={dispatch} />, { disableLifecycleMethods: true })
    wrapper.setState({ addingNewAddress: true })

    await wrapper.instance().nextSection()

    expect(createShippingAddressSpy).toHaveBeenCalledWith(checkout.shippingAddress)
    expect(setCartShippingAddressSpy).toHaveBeenCalledWith(20)
    expect(dispatch).toHaveBeenCalledWith('createShippingAddressAction')
    expect(dispatch).toHaveBeenCalledWith('setCartShippingAddressAction')
    expect(pushSpy).toHaveBeenCalledWith('/checkout/shipping-method')

    cookieSpy.mockRestore()
    pushSpy.mockRestore()
    createShippingAddressSpy.mockRestore()
    setCartShippingAddressSpy.mockRestore()
  })
})

describe('addressFormDisplayed()', () => {
  test('returns true when addressBook is empty', () => {
    const checkout = {
      addressBook: []
    }

    const wrapper = shallow(<ShippingAddressPage checkout={checkout} cart={{}} />, { disableLifecycleMethods: true })

    expect(wrapper.instance().addressFormDisplayed()).toBe(true)
  })

  test('returns true when adding new address is set in state', () => {
    const checkout = {
      addressBook: [{ id: 1 }]
    }

    const wrapper = shallow(<ShippingAddressPage checkout={checkout} cart={{}} />, { disableLifecycleMethods: true })
    wrapper.setState({ addingNewAddress: true })

    expect(wrapper.instance().addressFormDisplayed()).toBe(true)
  })

  test('returns true when address is not from the address book', () => {
    const checkout = {
      addressBook: [{ id: 1 }]
    }
    const cart = {
      shipping_address: {
        id: 2
      }
    }

    const wrapper = shallow(<ShippingAddressPage checkout={checkout} cart={cart} />, { disableLifecycleMethods: true })

    expect(wrapper.instance().addressFormDisplayed()).toBe(true)
  })
})

describe('cartAddressFromBook()', () => {
  test('returns false when customer is not logged in', () => {
    const checkout = {
      addressBook: []
    }

    const wrapper = shallow(<ShippingAddressPage cart={{}} checkout={checkout} />, { disableLifecycleMethods: true })

    expect(wrapper.instance().cartAddressFromBook()).toBe(false)
  })

  test("returns false when cart doesn't have a shipping address", () => {
    const cookieSpy = jest.spyOn(Cookies, 'get').mockImplementation(() => true)
    const checkout = {
      addressBook: []
    }

    const wrapper = shallow(<ShippingAddressPage cart={{}} checkout={checkout} />, { disableLifecycleMethods: true })

    expect(wrapper.instance().cartAddressFromBook()).toBe(false)

    cookieSpy.mockRestore()
  })

  test('returns false when shipping address is not from the address book', () => {
    const cookieSpy = jest.spyOn(Cookies, 'get').mockImplementation(() => true)
    const checkout = {
      addressBook: [{ id: 1 }]
    }
    const cart = {
      shipping_address: {
        id: 10
      }
    }

    const wrapper = shallow(<ShippingAddressPage cart={cart} checkout={checkout} />, { disableLifecycleMethods: true })

    expect(wrapper.instance().cartAddressFromBook()).toBe(false)

    cookieSpy.mockRestore()
  })

  test('returns true when shipping address is from the address book', () => {
    const cookieSpy = jest.spyOn(Cookies, 'get').mockImplementation(() => true)
    const checkout = {
      addressBook: [{ id: 10 }]
    }
    const cart = {
      shipping_address: {
        id: 10
      }
    }

    const wrapper = shallow(<ShippingAddressPage cart={cart} checkout={checkout} />, { disableLifecycleMethods: true })

    expect(wrapper.instance().cartAddressFromBook()).toBe(true)

    cookieSpy.mockRestore()
  })
})

test('renders address form components', () => {
  const checkout = {
    addressBook: [],
    shippingAddress: {
      errors: {}
    }
  }

  const cart = {
    line_items_count: 1
  }

  const wrapper = shallow(<ShippingAddressPage cart={cart} checkout={checkout} />, { disableLifecycleMethods: true })
  wrapper.setState({ loading: false })

  expect(wrapper).toMatchSnapshot()
  expect(wrapper.find(AddressFormHeader).length).toEqual(1)
  expect(wrapper.find(Button).length).toEqual(1)
  expect(wrapper.find(CheckoutAddressForm).length).toEqual(1)
})
