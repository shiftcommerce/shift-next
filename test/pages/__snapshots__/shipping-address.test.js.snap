// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders address form components 1`] = `
<div
  className="c-checkout__addressform"
>
  <div
    className="o-form__address"
  >
    <address_form_header_AddressFormHeader
      title="Shipping Address"
    />
    <AddressForm
      autoFillAddress={[Function]}
      checkout={
        Object {
          "addressBook": Array [],
          "shippingAddress": Object {
            "errors": Object {},
          },
        }
      }
      countries={
        Array [
          Object {
            "id": 1,
            "title": "United Kingdom",
            "value": "GB",
          },
          Object {
            "id": 2,
            "title": "United States",
            "value": "US",
          },
        ]
      }
      formName="shippingAddress"
      onBlur={[Function]}
      onChange={[Function]}
      onShowField={[Function]}
    />
    <div
      className="o-form__input-group"
    >
      <Button
        aria-label="View Shipping Options"
        className="c-address-form__button o-button--sml"
        disabled={true}
        label="View Shipping Options"
        onClick={[Function]}
        status="disabled"
        type="primary"
      />
    </div>
  </div>
</div>
`;
